src/tagmanager/tm_query.c

```diff
@@ -0,0 +1,299 @@
+ *
+ * @param q The query to operate on.
+ * @param lang The programming language, see @a TMParserType.
+ *
+ * @return 0 on succcess, < 0 on error.
+ */
+GEANY_API_SYMBOL
+gint tm_query_add_lang(TMQuery *q, TMParserType lang)
+{
+	g_array_append_val(q->langs, lang);
+
+	return 0;
+}
+
+
+/** Set the tag type filter of a query.
+ *
+ * The query results will be restricted to tags match any of the given types.
+ * You can only set this filter because the @a type is a bitmask containing
+ * multiple values.
+ *
+ * @param q The query to operate on.
+ * @param type Bitmask of the tag type to filter, see @a TMTagType.
+ *
+ * @return 0 on succcess, < 0 on error.
+ */
+GEANY_API_SYMBOL
+gint tm_query_set_tag_type(TMQuery *q, TMTagType type)
+{
+	if (type > tm_tag_max_t)
+		return -1;
+
+	q->type = type;
+
+	return 0;
+}
+
+
+/** Execute a query
+ *
+ * The query will be executed according to the applied filters. Filters
+ * of the same type are combined using logical OR. Then, filters of different
+ * types are combined using logcal AND.
+ *
+ * The execution is stateless. Filters can be added after execution and the
+ * query can be re-run. Previous runs don't add side-effects.
+ *
+ * The results are generally sorted by tag name, even if @a sort_attr is NULL.
+ * supplied. The results can be
+ * sorted by @a sort_attr, and deduplicated by @a dedup_attr. Deduplication
+ * happens after sorting. There is an additional limitation for deduplication:
+ * Only adjacent tags can be deduplicated so they must be suitably ordered.
+ * This can be achieved by using the same attributes for @a sort_attr and
+ * @a dedup_attr.
+ *
+ * The results are stored in a @c GPtrArray. The individual tags are still
+ * owned by the workspace but the container array must be freed with
+ * g_ptr_array_free().
+ *
+ * @param q The query to execute
+ * @param sort_attr @nullable @array{zero-terminated=1} Tag attribute to use
+ *                   as sort key, or @c NULL to not sort.
+ * @param dedup_attr @nullable @array{zero-terminated=1} Tag attribute to use
+ *                    as deduplicaiton key, or @c NULL to not perform deduplicaiton.
+ * @return @transfer{container} @elementtype{TMTag} Resulting tag array.
+ */
+GEANY_API_SYMBOL
+GPtrArray *tm_query_exec(TMQuery *q, TMTagAttrType *sort_attr, TMTagAttrType *dedup_attr)
+{
+	TMTag *tag;
+	guint i, ntags;
+	GString *s;
+	TMParserType *lang;
+	GPtrArray *ret;
+	TMTagAttrType def_sort_attr[] = { tm_tag_attr_name_t, 0 };
+
+	ret = g_ptr_array_new();
+
+	foreach_ptr_array(s, i, q->names)
+	{
+		TMTag **tags;
+		if (q->data_sources & TM_QUERY_SOURCE_GLOBAL_TAGS)
+		{
+			tags = tm_tags_find(q->workspace->global_tags, s->str, s->len, &ntags);
+			if (ntags)
+			{
+				g_ptr_array_set_size(ret, ret->len + ntags);
+				memcpy(&ret->pdata[ret->len], *tags, ntags * sizeof(gpointer));
+			}
+		}
+		if (q->data_sources & TM_QUERY_SOURCE_SESSION_TAGS)
+		{
+			tags = tm_tags_find(q->workspace->global_tags, s->str, s->len, &ntags);
```

Comment:
Is `q->workspace->global_tags` correct here? You do the same as in the TM_QUERY_SOURCE_GLOBAL_TAGS case above.